### PR TITLE
refactor(mempool)!: optimisations,excess sig index,fix weight calc

### PR DESCRIPTION
**Pull request by sdbondi**
Description
---

- fix weighting calculation for metadata (integer division bug: `(∑v_i) / d != v_1 / d + v_2/d +.... v_n/d` unless all `v`s are multiples of `d` due to integer rounding)
- updated fee estimation in wallet
- add excess signature index in mempool (required for compact block propagation)
- account for multiple kernels in transactions in mempool
- many smaller mempool optimisations (clones, etc)
- remove blocking thread pool usage for mempool
- remove unused mempool event stream (fills up and never empties, consuming some memory)
- clean up some legacy code (from when base node contained the wallet and miner)
- updated tests

Because the block weight calculation changed, this is a breaking change only affecting full/near full blocks

Motivation and Context
---
A bug in the way metadata weights are calculated meant that a set of transactions can have a different total weight from a block containing those same transactions. After a correctly weighted full block is returned from get block template, the block validator could incorrectly reject a block once mined as exceeding maximum weight.

Create a many to many index between transactions and kernels in the mempool.

How Has This Been Tested?
---
Unit: Existing tests updated
Manual: Producing 3 blocks worth of transactions, mining and checking the block is accepted and the transactions removed from the mempool

### CODE REVIEW

File: base_layer/core/src/mempool/priority/prioritized_transaction.rs [hunk 1]
```diff
@@ -20,70 +20,84 @@
...
-        Ok(Self {
-            priority: FeePriority::try_construct(&transaction, weight)?,
+        Self {
+            key,
+            priority: FeePriority::new(&transaction, weight),
             weight,
             transaction,
-            depended_output_hashes,
-        })
+            dependant_output_hashes: dependent_outputs.unwrap_or_default(),
```
Comment (delta1) [inline base_layer/core/src/mempool/priority/prioritized_transaction.rs]:
```suggestion
            dependent_output_hashes: dependent_outputs.unwrap_or_default(),
```

File: base_layer/core/src/mempool/service/service.rs [hunk 2]
```diff
@@ -238,33 +238,33 @@ impl MempoolService {
         });
     }
 
-    fn spawn_handle_incoming_request(&self, domain_msg: DomainMessage<mempool_proto::MempoolServiceRequest>) {
+    async fn spawn_handle_incoming_request(&self, domain_msg: DomainMessage<mempool_proto::MempoolServiceRequest>) {
         let inbound_handlers = self.inbound_handlers.clone();
         let outbound_message_service = self.outbound_message_service.clone();
-        task::spawn(async move {
-            let result = handle_incoming_request(inbound_handlers, outbound_message_service, domain_msg).await;
+        // task::spawn(async move {
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 3]
```diff
@@ -238,33 +238,33 @@ impl MempoolService {
...
+        let result = handle_incoming_request(inbound_handlers, outbound_message_service, domain_msg).await;
 
-            if let Err(e) = result {
-                error!(target: LOG_TARGET, "Failed to handle incoming request message: {:?}", e);
-            }
-        });
+        if let Err(e) = result {
+            error!(target: LOG_TARGET, "Failed to handle incoming request message: {:?}", e);
+        }
+        // });
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 4]
```diff
@@ -238,33 +238,33 @@ impl MempoolService {
...
     }
 
-    fn spawn_handle_incoming_response(&self, domain_msg: DomainMessage<mempool_proto::MempoolServiceResponse>) {
+    async fn spawn_handle_incoming_response(&self, domain_msg: DomainMessage<mempool_proto::MempoolServiceResponse>) {
         let waiting_requests = self.waiting_requests.clone();
-        task::spawn(async move {
-            let result = handle_incoming_response(waiting_requests, domain_msg.into_inner()).await;
+        // task::spawn(async move {
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 5]
```diff
@@ -238,33 +238,33 @@ impl MempoolService {
...
-                );
-            }
-        });
+        if let Err(e) = result {
+            error!(
+                target: LOG_TARGET,
+                "Failed to handle incoming response message: {:?}", e
+            );
+        }
+        // });
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 6]
```diff
@@ -289,43 +289,43 @@ impl MempoolService {
...
-    fn spawn_handle_local_request(
+    async fn spawn_handle_local_request(
         &self,
         request_context: RequestContext<MempoolRequest, Result<MempoolResponse, MempoolServiceError>>,
     ) {
         let mut inbound_handlers = self.inbound_handlers.clone();
-        task::spawn(async move {
-            let (request, reply_tx) = request_context.split();
-            let result = reply_tx.send(inbound_handlers.handle_request(request).await);
+        // task::spawn(async move {
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 7]
```diff
@@ -289,43 +289,43 @@ impl MempoolService {
...
-    fn spawn_handle_block_event(&self, block_event: Arc<BlockEvent>) {
+    async fn spawn_handle_block_event(&self, block_event: Arc<BlockEvent>) {
         let mut inbound_handlers = self.inbound_handlers.clone();
-        task::spawn(async move {
-            let result = inbound_handlers.handle_block_event(&block_event).await;
-            if let Err(e) = result {
-                error!(target: LOG_TARGET, "Failed to handle base node block event: {:?}", e);
-            }
-        });
+        // task::spawn(async move {
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

File: base_layer/core/src/mempool/service/service.rs [hunk 8]
```diff
@@ -289,43 +289,43 @@ impl MempoolService {
...
+        }
+        // });
     }
 
-    fn spawn_handle_request_timeout(&self, timeout_request_key: u64) {
+    async fn spawn_handle_request_timeout(&self, timeout_request_key: u64) {
         let waiting_requests = self.waiting_requests.clone();
-        task::spawn(async move {
-            let result = handle_request_timeout(waiting_requests, timeout_request_key).await;
+        // task::spawn(async move {
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/service/service.rs]: remove?

**Review by SWvheerden — COMMENTED** [review]
LGTM
There is just some code that's been commented out that's been left in.

File: base_layer/core/src/mempool/service/service.rs (hunk 2 above)
Comment (sdbondi) [reply base_layer/core/src/mempool/service/service.rs]: Thanks -  I think I'll bring them back in and leave it for another refactor to evaluate where we can cut down on these spawns

**Review by stringhandler — APPROVED** [review]
looking good.

**Pull request merged**